backend-lib/src/main/rust/lib.rs

```diff
@@ -2368,6 +2371,138 @@ pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustDerivationTool_de
+        &[
+            (&env.byte_array_from_slice(key.data())?).into(),
+            (&env.byte_array_from_slice(key.chain_code().as_bytes())?).into(),
+        ],
+    )?)
+}
+
+#[unsafe(no_mangle)]
+pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustDerivationTool_deriveAccountMetadataKeyFromSeed<
+    'local,
+>(
+    mut env: JNIEnv<'local>,
+    _: JClass<'local>,
+    seed: JByteArray<'local>,
+    account_index: jlong,
+    network_id: jint,
+) -> jobject {
+    let res = catch_unwind(&mut env, |env| {
+        let _span =
+            tracing::info_span!("RustDerivationTool.deriveAccountMetadataKeyFromSeed").entered();
+        let network = parse_network(network_id as u32)?;
+        let seed = secret_from_jni(env, seed)?;
+        let account = zip32_account_index_from_jlong(account_index)?;
+
+        let key = zip32::registered::SecretKey::from_subpath(
+            b"MetadataKeys",
+            seed.expose_secret(),
+            // TODO: Change this to whatever ZIP number is assigned to the metadata key ZIP draft.
+            325,
+            &[
+                PathElement::new(ChildIndex::hardened(network.coin_type()), &[]),
+                PathElement::new(ChildIndex::hardened(account.into()), &[]),
+            ],
+        )?;
+
+        Ok(encode_metadata_key(env, key)?.into_raw())
+    });
+    unwrap_exc_or(&mut env, res, ptr::null_mut())
+}
+
+#[unsafe(no_mangle)]
+pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustDerivationTool_derivePrivateUseMetadataKey<
+    'local,
+>(
+    mut env: JNIEnv<'local>,
+    _: JClass<'local>,
+    account_metadata_key_sk: JByteArray<'local>,
+    account_metadata_key_c: JByteArray<'local>,
+    ufvk_string: JString<'local>,
+    private_use_subject: JByteArray<'local>,
+    network_id: jint,
+) -> jobjectArray {
+    let res = catch_unwind(&mut env, |env| {
+        let _span = tracing::info_span!("RustDerivationTool.derivePrivateUseMetadataKey").entered();
+        let account_metadata_key_sk = utils::java_bytes_to_rust(env, &account_metadata_key_sk)?;
+        let account_metadata_key_c = utils::java_bytes_to_rust(env, &account_metadata_key_c)?;
+        let ufvk_string = utils::java_nullable_string_to_rust(env, &ufvk_string)?;
+        let private_use_subject = utils::java_bytes_to_rust(env, &private_use_subject)?;
+        let network = parse_network(network_id as u32)?;
+
+        let account_metadata_key = {
+            let sk = account_metadata_key_sk
+                .as_slice()
+                .try_into()
+                .map_err(|_| anyhow!("Incorrect length for account_metadata_key_sk"))?;
+
+            let chain_code = ChainCode::new(
+                account_metadata_key_c
+                    .as_slice()
+                    .try_into()
+                    .map_err(|_| anyhow!("Incorrect length for account_metadata_key_c"))?,
+            );
+
+            zip32::registered::SecretKey::from_parts(sk, chain_code)
+        };
```

Comment:
I still think the metadata key parsing would be better split out into a helper, but :shrug:.